[WIP] Refactor/chat views and subs

yenda opened the pull request:
[WIP] cutting the PR in smaller pieces

PR is long because:
- most of chat.subs and contact.subs code is moved to chat.db and contact.db
- utils.contacts is moved to contact.db
- lots of logic is moved from views to subs
- tests are touched

replaces #6652 partially (block feature will be added separately to reduce size of PR)
fix #6033 remove logic from view

- [fix] usernames in bold font on Linux
- [fix] current chat in bold font Linux
- [refactor] remove defview and subscriptions to
use pure functions with params, use flatlist on
desktop for chats like it is done in mobile
- [refactor] optimize subscriptions for perf, remove
message-groups
- [fix] prevent use of rtl char in username to impersonate
other other user random name

### Summary:

Simplify views and subscriptions to facilitate implementation of block contact feature and other future improvements.

Only root views are subscribing and using `defview` macro, children use reagent form-1 components as much as possible.
Priority map is used to keep messages of chats in order without having to sort them every time so that only a single reduce pass is needed to enrich them with contextual data.

#### Platforms
- Android
- iOS
- macOS
- Linux
- Windows

#### Areas that maybe impacted
**Functional**
- 1-1 chats
- public chats
- group chats

**Non-functional**
- battery performance
- CPU performance / speed of the app
- network consumption

status: wip (still have to check if mobile works)

Commit:
[refactor] chat views and subscriptions

- [fix] usernames in bold font on Linux
- [fix] current chat in bold font Linux
- [refactor] remove defview and subscriptions to
use pure functions with params, use flatlist on
desktop for chats like it is done in mobile
- [refactor] optimize subscriptions for perf, remove
message-groups
- [fix] prevent use of rtl char in username to impersonate
other other user random name

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

rasom commented:
@yenda is it possible to implement #6652 without moving code from/to different namespaces? Has it been discussed?

rasom commented on src/status_im/utils/priority_map.cljs:
```diff
@@ -0,0 +1,229 @@
+(ns status-im.utils.priority-map
```
@yenda is it copied from somewhere or that's yours implementation?

yenda replied:
yes it's https://github.com/tailrecursion/cljs-priority-map/blob/master/src/cljs/tailrecursion/priority_map.cljs with a bug fixv for keys and val fns

yenda replied:
do you think I should clone the repo instead @rasom ?

rcullito commented on src/status_im/chat/models/loading.cljs:
```diff
@@ -70,17 +39,17 @@
               {:db (assoc db
                           :chats          chats
                           :contacts/dapps default-dapps)}
-              (group-messages)
```
I know you mention it in your PR summary, but why are these getting removed? group messages are no longer being supported across the app?

yenda replied:
this is not group chat, it was grouping of messages by datemark, done at message reception for caching, but now the subscriptions are more efficient and messages need to be filtered alter for tag features so the cache will too often be invalid to be useful

rcullito replied:
thanks for clarifying.

rcullito commented on src/status_im/constants.cljs:
```diff
@@ -18,7 +18,7 @@
 (def max-chat-name-length 20)
 (def response-suggesstion-resize-duration 100)
-(def default-number-of-messages 20)
+(def default-number-of-messages 100)
```
this will be # of messages loaded initially?

yenda replied:
per chat, but good catch I'll try less 100 might be too much it was to smooth scrolling

rcullito commented on src/status_im/ui/screens/chat/message/message.cljs:
```diff
@@ -41,22 +37,22 @@
-     [react/text {:style           (style/quoted-message-text outgoing)
-                  :number-of-lines 5}
-      text]]))
+(defview quoted-message
```
can this become a pure fn or does it need to stay as a defview?

yenda replied:
should be a defn thanks

rcullito commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -1,143 +1,139 @@
+                         :handler   #(on-options chat-id name group-chat public?)}]])]
+   (when-not (or public? group-chat) [add-contact-bar contact])])
 
-(defmulti message-row (fn [{{:keys [type]} :row}] type))
```
any conceptual reason behind this change from *message-row* to *message-view*?

yenda replied:
just to align with mobile

rcullito replied:
got it. thanks.

rcullito commented on src/status_im/ui/screens/desktop/main/chat/styles.cljs:
```diff
@@ -38,6 +38,7 @@
 
 (def author
   {:color         colors/black
+   :font-weight   :bold
```
:tada: :balloon:

rcullito reviewed (approved):
big overhaul! your changes definitely improved readability for someone not super familiar with the codebase. a couple of questions inline on concepts and various naming conventions.  :+1:

oskarth commented:
I find it difficult to believe that such a big PR won't have a bunch of unforseen regressions. Can we make this diff smaller and just address minimal necessary to solve the problem at hand, instead of cramming a bunch of stuff together?

`git add -p` is your friend!

Commit:
[refactor] move contact.subs and utils.contacts to contact.db

functions defined in subs need to be reusable and testable independently
from re-frame framework. they are moved into db namespace for that purpose

lukaszfryc commented:
@churik could you help me with testing this PR? Especially on Linux and Windows. I can take mobile + macOS. 

> [fix] usernames in bold font on Linux
[fix] current chat in bold font Linux

churik commented:
@lukaszfryc I'm not sure it is ready according to https://github.com/status-im/status-react/pull/6764#issuecomment-439765593 and Roman's comments.
WDYT?

lukaszfryc commented:
@yenda, should we wait with testing?

yenda commented:
@lukaszfryc yes please wait

lukaszfryc commented:
@yenda, any update?

yenda commented:
@lukaszfryc I will close this PR once I'm done moving all the changes to separate PRs

ghost commented:
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

ghost commented:
This issue has been automatically closed. Please re-open if this issue is important to you.
